controllers/flow_tree.go

```diff
@@ -0,0 +1,326 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	ytv1 "github.com/ytsaurus/ytsaurus-k8s-operator/api/v1"
+	apiProxy "github.com/ytsaurus/ytsaurus-k8s-operator/pkg/apiproxy"
+	"github.com/ytsaurus/ytsaurus-k8s-operator/pkg/consts"
+)
+
+type (
+	stepResultMark string
+)
+
+const (
+	stepResultMarkUnsatisfied stepResultMark = ""
+	stepResultMarkHappy       stepResultMark = "happy"
+	stepResultMarkUnhappy     stepResultMark = "unhappy"
+)
+
+var terminateTransitions = map[ytv1.UpdateState]ytv1.ClusterState{
+	ytv1.UpdateStateImpossibleToStart: ytv1.ClusterStateCancelUpdate,
+}
+
+type flowCondition func(ctx context.Context, ytsaurus *apiProxy.Ytsaurus, componentManager *ComponentManager) stepResultMark
+
+func flowCheckStatusCondition(conditionName string) flowCondition {
+	return func(ctx context.Context, ytsaurus *apiProxy.Ytsaurus, componentManager *ComponentManager) stepResultMark {
+		if ytsaurus.IsUpdateStatusConditionTrue(conditionName) {
+			return stepResultMarkHappy
+		}
+		return stepResultMarkUnsatisfied
+	}
+
+}
+
+type flowStep struct {
+	updateState ytv1.UpdateState
+	// For most of the steps, there will be only one next step,
+	// but for some outcome is based on condition result.
+	nextSteps map[stepResultMark]*flowStep
+}
+
+func newSimpleStep(updateState ytv1.UpdateState) *flowStep {
+	return &flowStep{
+		updateState: updateState,
+	}
+}
+
+func newConditionalForkStep(updateState ytv1.UpdateState, unhappyNext *flowStep) *flowStep {
+	return &flowStep{
+		updateState: updateState,
+		nextSteps: map[stepResultMark]*flowStep{
+			stepResultMarkUnhappy: unhappyNext,
+		},
+	}
+}
+
+func (s *flowStep) checkCondition(ctx context.Context, ytsaurus *apiProxy.Ytsaurus, componentManager *ComponentManager) stepResultMark {
+	condition := flowConditions[s.updateState]
+	return condition(ctx, ytsaurus, componentManager)
+}
+
+func (s *flowStep) next(mark stepResultMark) *flowStep {
+	if mark == stepResultMarkUnsatisfied {
+		return nil
+	}
+	return s.nextSteps[mark]
+}
+
+func (s *flowStep) addHappyPath(next *flowStep) *flowStep {
+	s.nextSteps = map[stepResultMark]*flowStep{
+		stepResultMarkHappy: next,
+	}
+	return next
+}
+
+func (s *flowStep) chain(steps ...*flowStep) *flowStep {
+	current := s
+	for _, step := range steps {
+		current.addHappyPath(step)
+		current = step
+	}
+	return current
+}
+
+// for the fast lookup of the step by the update state
+// otherwise we can traverse the graph each time
+type flowTree struct {
+	index map[ytv1.UpdateState]*flowStep
+	head  *flowStep
+	tail  *flowStep
```

Comment:
We are storing map state to step (as in original PR) and also have directed graph for steps, which gives us the possibility not to declare next step when adding one.